### PR TITLE
[HW] Lower hw.instance_choice to SV

### PR DESCRIPTION
**Pull request by nandor**
This PR sketches the lowering of instance choices to SystemVerilog, emitting the placeholder macros and producing the output for the default branches.

The emission of the alternatives will be done in a follow-up PR. 

An integration test is also built out to verify the correctness of the behaviour of the ops and to verify the correctness of the file list and macro definition groups involved.

SV code is emitted directly instead of building IR for it at the moment as regular instance operations cannot use a symbolic (macro) reference to name the target module.

Upon further thought, the details of the lowerings will not be included in the FIRRTL ABI: the ABI will pin down the ordering in which files emitted by the compiler will be consumed, however their contents which implement instance choices will be left up to the implementation, i.e. CIRCT.

### CODE REVIEW

File: integration_test/EmitVerilog/instance-choice.mlir.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+//===- driver.cpp - Verilator software driver -----------------------------===//
```
Comment (teqdruid) [inline integration_test/EmitVerilog/instance-choice.mlir.cpp:1]: Why did you fork this driver code?

File: integration_test/EmitVerilog/instance-choice.mlir [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// REQUIRES: verilator
+// RUN: circt-opt %s -export-split-verilog='dir-name=%t.dir' -o /dev/null
```
Comment (teqdruid) [inline integration_test/EmitVerilog/instance-choice.mlir:2]: Why aren't you using circt-rtl-sim?

Comment (nandor) [reply integration_test/EmitVerilog/instance-choice.mlir:2]: The tool makes too many assumptions about the source file list. Here we want to pass in a bunch of addition `-F` files and options in order since we are testing the ABI against a specific tool.

File: integration_test/EmitVerilog/instance-choice.mlir.cpp (hunk 1 above)
Comment (nandor) [reply integration_test/EmitVerilog/instance-choice.mlir.cpp]: Manged to re-use it.

File: integration_test/EmitVerilog/instance-choice.mlir [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+// RUN: verilator %driver --cc --sv --exe --build -I%t.dir -F %t.dir/filelist.f -o %t.exe --top-module top
+// RUN: %t.exe --cycles 10 2>&1 | FileCheck %s --check-prefix=DEFAULT
```
Comment (seldridge) [inline integration_test/EmitVerilog/instance-choice.mlir:4]: Does this need a full-blow test or is `verilator --lint-only` sufficient? It seems like no. Given that the alternative emission isn't yet implemented, the check can be as simple as putting an `assert(false)` in the module you want to make sure doesn't run.

Nit: You may be able to avoid file output by piping non-split output directly to Verilator. It will take `circt-opt ... | verilator /dev/stdin -top-module top`. The nit is only for consideration.

File: integration_test/lit.cfg.py [hunk 4]
```diff
@@ -206,6 +206,7 @@
   config.available_features.add('circt-lec')
   tools.append('circt-lec')
 
+config.substitutions.append(('%driver', f'{config.driver}'))
```
Comment (seldridge) [inline integration_test/lit.cfg.py:209]: Is this and the other lit change necessary given that this is hand-rolling a verilator call?

File: integration_test/EmitVerilog/instance-choice.mlir (hunk 3 above)
Comment (seldridge) [reply integration_test/EmitVerilog/instance-choice.mlir:4]: Does this string work on Windows given the use of a forward slash? If not, is the test properly guarded so it doesn't run on the Windows runner? You can run the windows integration test manually from the `Actions` tab at the top.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 5]
```diff
@@ -3728,7 +3728,12 @@ class StmtEmitter : public EmitterBase,
   LogicalResult visitSV(AliasOp op);
   LogicalResult visitSV(InterfaceInstanceOp op);
   LogicalResult visitStmt(OutputOp op);
+
```
Comment (seldridge) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:3731]: Nit: Unrelated whitespace change.

File: test/Conversion/ExportVerilog/instance-choice.mlir [hunk 6]
```diff
@@ -0,0 +1,40 @@
+// RUN: circt-opt %s -export-split-verilog='dir-name=%t.dir'
+// RUN: cat %t.dir/top.sv | FileCheck %s
```
Comment (seldridge) [inline test/Conversion/ExportVerilog/instance-choice.mlir:2]: Single file emission would keep this to one line and avoid any output (or Windows issues).

File: test/Conversion/ExportVerilog/instance-choice.mlir [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+hw.module private @TargetDefault(in %a: i32, out b: i32) {
+  hw.output %a : i32
+}
+
+// CHECK-LABEL: module top
+hw.module public @top(in %a: i32, out b: i32, out d: i32) {
+  // CHECK:      `ifndef __circt_choice_top_inst1
+  // CHECK-NEXT: `define __circt_choice_top_inst1 TargetDefault
+  // CHECK-NEXT: `endif
+  // CHECK-NEXT: __circt_choice_top_inst1 inst1 (
```
Comment (seldridge) [inline test/Conversion/ExportVerilog/instance-choice.mlir:21]: Nit: shouldn't this be: 

```suggestion
  // CHECK-NEXT: `__circt_choice_top_inst1 inst1 (
```

Ditto below.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 8]
```diff
@@ -4986,9 +4991,71 @@ LogicalResult StmtEmitter::visitStmt(InstanceOp op) {
...
+}
+
+LogicalResult StmtEmitter::visitStmt(InstanceChoiceOp op) {
+  HWModuleOp parentMod = op->getParentOfType<hw::HWModuleOp>();
+
+  std::string referenceMacroName;
+  {
+    llvm::raw_string_ostream os(referenceMacroName);
+    os << "__circt_choice_" << getVerilogModuleName(parentMod) << "_"
+       << getSymOpName(op);
```
Comment (seldridge) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:5019]: How to handle name conflicts with this macro?

How about this:

The instance choice op needs to be given a macro decl that it is associated with. This is purely to guarantee a unique name and for no other reason. That is then substituted in here instead of generating a potential name conflict on the fly.

To check this, there should be a test where there is a pre-existing macro decl with this name and there is no conflict in emission.

File: test/Conversion/ExportVerilog/instance-choice.mlir (hunk 7 above)
Comment (seldridge) [reply test/Conversion/ExportVerilog/instance-choice.mlir:21]: The code is already generating the `` ` ``, so everything should the check is just not strict enough!

**Review by seldridge — COMMENTED** [review]
This is looking great. Emission looks solid.

Only major concern is around macro conflicts. This can be fixed by making the instance choice a symbol user. WDYT?

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp (hunk 5 above)
Comment (nandor) [reply lib/Conversion/ExportVerilog/ExportVerilog.cpp:3731]: It is related, I want to group the calls and the helper they resort to.

File: integration_test/lit.cfg.py (hunk 4 above)
Comment (nandor) [reply integration_test/lit.cfg.py:209]: We must expose the path to the driver somehow and it's in a subdirectory far away.

File: integration_test/EmitVerilog/instance-choice.mlir (hunk 3 above)
Comment (nandor) [reply integration_test/EmitVerilog/instance-choice.mlir:4]: I do want an execution check here, the cost of it should be irrelevant. Replaced slashes with `%{fs-sep}`, that should work.

Comment (nandor) [reply integration_test/EmitVerilog/instance-choice.mlir:4]: Also, I do want this to become an integration test for multi-file emission, with verilator being told to ingest the correct file lists.

File: lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+} // end anonymous namespace
+
+LogicalResult
+ExportVerilog::lowerHWInstanceChoices(mlir::ModuleOp module,
+                                      const LoweringOptions &options) {
+  // Collect all instance choices & symbols.
+  SmallVector<InstanceChoiceOp> instances;
+  SymbolCache symCache;
+  {
```
Comment (seldridge) [inline lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp:58]: Interesting use of a scope here. Normally I expect this if a PR is trying to hide variables from the outer scope (as the PR does with `os` later which looks good!). What is this accomplishing here? Is this trying to say that everything in the scope is related to the initialization of the previous two variables?

File: lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    auto defaultModuleOp = cast<HWModuleLike>(
+        symCache.getDefinition(inst.getDefaultModuleNameAttr()));
+
+    // Generate a macro name to describe the target of this instance.
+    std::string name;
+    {
+      llvm::raw_string_ostream os(name);
+      os << "__circt_choice_" << parent.getName() << "_"
+         << inst.getInstanceName();
+    }
```
Comment (seldridge) [inline lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp:89]: As stated above, nice scope to hide `os` from unnecessarily polluting the outer namespace.

Comment (seldridge) [reply lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp:89]: That said, please have the PR favor `SmallString` as this is more aligned with the LLVM Developer Policy and would avoid a heap allocation as long as the string is below the pre-allocated size.

File: lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    // instance choices to a better SV-level representation of the operation.
+    ImplicitLocOpBuilder builder(inst.getLoc(), inst);
+    builder.create<sv::IfDefOp>(
+        symName, [&] {},
+        [&] {
+          builder.create<sv::MacroDefOp>(
+              symRef, builder.getStringAttr("{{0}}"),
+              builder.getArrayAttr(
+                  {FlatSymbolRefAttr::get(defaultModuleOp.getNameAttr())}));
+        });
```
Comment (seldridge) [inline lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp:110]: Consider that the PR can re-use the builder it already has, moving the insertion point, and resetting it with a guard. I think the following works. `declBuilder` is used for consistency, but this is really just `builder` with the patch below applied:

```suggestion
    OpBuilder::InsertionGuard guard(declBuilder);
    declBuilder.setInsertionPoint(inst);
    declBuilder.create<sv::IfDefOp>(inst.getLoc(),
        symName, [&] {},
        [&] {
          builder.create<sv::MacroDefOp>(
              symRef, builder.getStringAttr("{{0}}"),
              builder.getArrayAttr(
                  {FlatSymbolRefAttr::get(defaultModuleOp.getNameAttr())}));
        });
```

File: lib/LogicalEquivalence/LogicExporter.cpp [hunk 12]
```diff
@@ -87,6 +87,10 @@ struct Visitor : public hw::StmtVisitor<Visitor, LogicalResult>,
     return failure();
   }
 
+  LogicalResult visitStmt(hw::InstanceChoiceOp op) {
+    return op.emitError("instance choices are not supported");
+  }
```
Comment (seldridge) [inline lib/LogicalEquivalence/LogicExporter.cpp:92]: Thanks for adding this backstop error!

File: test/Conversion/ExportVerilog/hw-lower-instance-choices.mlir [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+hw.module private @TargetDefault(in %a: i32, out b: i32) {
+  hw.output %a : i32
+}
+
+hw.module public @top(in %a: i32, out b: i32, out d: i32) {
+  // CHECK:               sv.ifdef  "__circt_choice_top_inst1" {
+  // CHECK-NEXT:          } else {
+  // CHECK-NEXT{LITERAL}:   sv.macro.def @__circt_choice_top_inst1 "{{0}}"([@TargetDefault])
+  // CHECK-NEXT:          }
```
Comment (seldridge) [inline test/Conversion/ExportVerilog/hw-lower-instance-choices.mlir:22]: Aside: I haven't looked at this serialization format in a long time. This would be clearer if `sv.ifdef` serialized with an actual symbol and not a string.

File: test/Conversion/ExportVerilog/hw-lower-instance-choices.mlir [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  // CHECK-SAME:          {hw.choiceTarget = @__circt_choice_top_inst1}
+  %b = hw.instance_choice "inst1" sym @inst1 @TargetDefault or @TargetA if "A" or @TargetB if "B"(a: %a: i32) -> (b: i32)
+
+  // CHECK:               sv.ifdef  "__circt_choice_top_inst2" {
+  // CHECK-NEXT:          } else {
+  // CHECK-NEXT{LITERAL}:   sv.macro.def @__circt_choice_top_inst2 "{{0}}"([@TargetB])
+  // CHECK-NEXT:          }
+  // CHECK-NEXT:          hw.instance_choice "inst2"
+  // CHECK-SAME:          {hw.choiceTarget = @__circt_choice_top_inst2}
+  %c = hw.instance_choice "inst2" sym @inst2 @TargetB or @TargetA if "A" or @TargetDefault if "B"(a: %a: i32) -> (b: i32)
```
Comment (seldridge) [inline test/Conversion/ExportVerilog/hw-lower-instance-choices.mlir:33]: Nice clean test and clean presentation of FileCheck directives! This helps with maintainability going forward. 💯

File: include/circt/Conversion/Passes.td [hunk 15]
```diff
@@ -104,6 +104,20 @@ def LegalizeAnonEnums : Pass<"legalize-anon-enums", "mlir::ModuleOp"> {
   ];
 }
 
+def HWLowerInstanceChoices : Pass<"hw-lower-instance-choices",
+                                  "mlir::ModuleOp"> {
+  let summary = "Prepare the collateral for instance choice emission";
+  let description = [{
+    This pass runs as part of verilog emission.
+    It introduces the macros & file lists to which instance choices lower to.
```
Comment (seldridge) [inline include/circt/Conversion/Passes.td:112]: Nit: this doesn't introduce any file lists yet.

File: test/Conversion/ExportVerilog/instance-choice.mlir [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+// CHECK-LABEL: module top
+hw.module public @top(in %a: i32, out b: i32, out d: i32) {
+  // CHECK:      `ifndef __circt_choice_top_inst1
+  // CHECK-NEXT: `define __circt_choice_top_inst1 TargetDefault
+  // CHECK-NEXT: `endif
+  // CHECK-NEXT: `__circt_choice_top_inst1 inst1 (
+  // CHECK-NEXT:    .a (a),
+  // CHECK-NEXT:    .b (b)
+  // CHECK-NEXT: );
+  %b = hw.instance_choice "inst1" sym @inst1 @TargetDefault or @TargetA if "A" or @TargetB if "B"(a: %a: i32) -> (b: i32)
```
Comment (seldridge) [inline test/Conversion/ExportVerilog/instance-choice.mlir:25]: Emission looks great. Restating my comment above that this test is both very clean and well formatted for both understandability and maintainability.

**Review by seldridge — APPROVED** [review]
Looks good! Thanks for iterating towards the `sv.decl` solution.

Cleanup the comments and this is good to go.

It's not ideal that the PR is relying on the order of an `ifdef` op before an instance choice op. This is less of a concern if this is hidden behind the prepare pipeline (which can make assumptions that the IR will be unmodified because it is supposed to be run as a unit). This is more of a concern if this is a real, standalone pass that is supposed to be run anytime.

File: lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/LLVM.h"
+#include "circt/Support/Namespace.h"
+#include "circt/Support/SymCache.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/Support/Debug.h"
+
+#include <mutex>
```
[the remark this answers is not in the document]
Comment (nandor) [reply lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp:33]: Neat!

File: lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp (hunk 9 above)
Comment (nandor) [reply lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp:58]: Useless now, but if there are multiple constructs, I do try to isolate them under the shared data they build.

File: test/Conversion/ExportVerilog/hw-lower-instance-choices.mlir (hunk 13 above)
Comment (nandor) [reply test/Conversion/ExportVerilog/hw-lower-instance-choices.mlir:22]: We have a bit of a problem here, we'd need macro-level expressions.
